### PR TITLE
doc: add page about creating an application

### PR DESCRIPTION
**Pull request by authmillenon**
Another one in #4309 bites the dust ;-)

### CODE REVIEW

**Comment by authmillenon** [comment]
Or maybe @kaspar030 want's to review?

File: doc/doxygen/src/creating-an-application.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+Creating an application                              {#creating-an-application}
+=======================
+
+[TOC]
+
+To create your own application you need to create a directory contating some 
+from the C file(s) containing your source code and a Makefile. A template
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: "... some of ...", but I'd drop that and just write "you need to create a directory contating the C file(s) containing your source code and a Makefile"

Also "s/contaiting/containing".

File: doc/doxygen/src/creating-an-application.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+[RIOT repository](https://github.com/RIOT-OS/RIOT).
+
+The main function                                          {#the-main-function}
+=================
+After the board is initialized, RIOT starts two threads: the idle thread and 
+the main thread. The idle thread has the lowest priority and will run whenever 
+no other thread is ready to run. It will automatically use the lowest possible 
+power mode for the device. The main thread - configured with a default priority 
+that is right in the middle between the lowest and the highest available 
+priority - is the first thread that runs and calls the `main()` function. This
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: s/priority/priorities/

File: doc/doxygen/src/creating-an-application.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    return 0;
+}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The application's Makefile                         {#the-applications-makefile}
+==========================
+The minimal Makefile                                    {#the-minimal-makefile}
+--------------------
+At minimum the Makefile of an application (see #the-build-system) needs the 
+following macros defined:
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: macros -> variables?

File: doc/doxygen/src/creating-an-application.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ * `APPLICATION`: should contain the name of your application
+ * `RIOTBASE`: specifies the path to your copy of the RIOT repository (note 
+   that you may want to use `$(CURDIR)` here, to give a relative path)
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: maybe clarify to use ```$(CURDIR)/../.." or something?

File: doc/doxygen/src/creating-an-application.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+The `BOARD` macro is also required and recommended to be set to `native` per 
+default, but is recommended to be overwritable with the `?=` operate.
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: overwritable -> overridable, operate -> operator

File: doc/doxygen/src/creating-an-application.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.mk}
+# Set the name of your application:
+APPLICATION = foobar
+
+# If no BOARD is found in the environment, use this default:
+BOARD ?= native
+
+# This has to be the absolute path to the RIOT base directory:
+RIOTBASE ?= $(CURDIR)/../../RIOT
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: maybe clear enough. ;)

File: doc/doxygen/src/creating-an-application.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+`BOARD_WHITELIST` will issue a message that one has to expect errors if they 
+build the application for the board referred by that name and the list can be 
+used by a CI system to not build the application for this board at all. A board 
+that is included in `BOARD_BLACKLIST` will show the same behavior. The build 
+system evaluates `BOARD_WHITELIST` first and then `BOARD_BLACKLIST`. The
+`BOARD_INSUFFICIENT_MEMORY` macro is similar to `BOARD_BLACKLIST`, but will 
+build in any case. A CI system can use the information provided by the 
+`BOARD_INSUFFICIENT_MEMORY` macro to skip the linking step in the build 
+process, since some linkers will issue an error when the code won't fit the 
+target board's flash memory.
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: mention too much RAM usage?

File: doc/doxygen/src/creating-an-application.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+Including modules                                          {#including-modules}
+-----------------
+By default a RIOT application comprises only the applications' code itself, the
```
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: comprises only *of*

Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: sorry, not sure here

**Comment by kaspar030** [comment]
I'm on arch, so I cannot test doxygen. Text looks fine!

File: doc/doxygen/src/creating-an-application.md (hunk 3 above)
Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: Depending on the version of make [macro and variable are used interchangeably](http://www.gnu.org/software/make/manual/make.html#Using-Variables)

File: doc/doxygen/src/creating-an-application.md (hunk 4 above)
Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: That's even more confusing IMHO.

Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: The example might give an impression what is meant

Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: yep, saw it later. fine like this.

File: doc/doxygen/src/creating-an-application.md (hunk 3 above)
Comment (kaspar030) [inline doc/doxygen/src/creating-an-application.md]: ok

**Comment by authmillenon** [comment]
In pure markdown it should work. The bug only is apparent with block-comments:

    ~~~~~~~~
    some code
    ~~~~~~~~

vs

    /**
     * ~~~~~~~
     * some code
     * ~~~~~~~
     */

File: doc/doxygen/src/creating-an-application.md (hunk 4 above)
Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: Sorry, did see your comment in the example only now ;-)

**Comment by authmillenon** [comment]
Addressed comments.

**Comment by authmillenon** [comment]
Addressed comments

File: doc/doxygen/src/creating-an-application.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+The minimal Makefile                                    {#the-minimal-makefile}
+--------------------
+At minimum the Makefile of an application (see @ref getting-started) needs the
+following macros defined:
+ * `APPLICATION`: should contain the name of your application
+ * `RIOTBASE`: specifies the path to your copy of the RIOT repository (note 
+   that you may want to use `$(CURDIR)` here, to give a relative path)
+
+The `BOARD` macro is also required and recommended to be set to `native` per 
+default, but is recommended to be overridable with the `?=` operator.
```
Comment (Lotterleben) [inline doc/doxygen/src/creating-an-application.md]: nitpick: I think “by default” reads more smoothly :D

**Comment by Lotterleben** [comment]
Just plopping by to say I think this is awesome. <3

File: doc/doxygen/src/creating-an-application.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+To create your own application you need to create a directory containing one or
+multiple C file(s) with your source code and a Makefile. A template Makefile is
+available in the `dist` folder of the
+[RIOT repository](https://github.com/RIOT-OS/RIOT).
+
+The main function                                          {#the-main-function}
+=================
+After the board is initialized, RIOT starts two threads: the idle thread and 
+the main thread. The idle thread has the lowest priority and will run whenever 
+no other thread is ready to run. It will automatically use the lowest possible
```
Comment (PeterKietzmann) [inline doc/doxygen/src/creating-an-application.md]: Well, that's the plan but not really the fact I guess :/

File: doc/doxygen/src/creating-an-application.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    return 0;
+}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The application's Makefile                         {#the-applications-makefile}
+==========================
+The minimal Makefile                                    {#the-minimal-makefile}
+--------------------
+At minimum the Makefile of an application (see @ref getting-started) needs the
+following macros defined:
```
Comment (PeterKietzmann) [inline doc/doxygen/src/creating-an-application.md]: "Needs to define..." as an alternative

File: doc/doxygen/src/creating-an-application.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+How to handle unsupported boards?                  {#handle-unsupported-boards}
+---------------------------------
+Sometimes it is necessary to exclude boards because they don't provide a 
+required functionality or don't have sufficient memory. RIOT's build system 
+looks for the macros `BOARD_BLACKLIST`, `BOARD_WHITELIST`, and 
+`BOARD_INSUFFICIENT_MEMORY`. Any board name that is not included in 
+`BOARD_WHITELIST` will issue a message that one has to expect errors if they 
+build the application for the board referred by that name and the list can be 
+used by a CI system to not build the application for this board at all. A board 
+that is included in `BOARD_BLACKLIST` will show the same behavior. The build
```
Comment (PeterKietzmann) [inline doc/doxygen/src/creating-an-application.md]: Maybe you can split that sentence into two

**Comment by PeterKietzmann** [comment]
Again: Thanks Martine for taking the initiative :-) !

File: doc/doxygen/src/creating-an-application.md (hunk 10 above)
Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: "Documentation is also specification, not just fact stating" I counter to that ^^

File: doc/doxygen/src/creating-an-application.md (hunk 12 above)
Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: Wrong line? Both sentences in this line are sufficiently short and I don't know how to split them more.

Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: I assume you mean the sentence two lines above.... Will fix.

File: doc/doxygen/src/creating-an-application.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    puts("Hello World!");
+    return 0;
+}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The application's Makefile                         {#the-applications-makefile}
+==========================
+The minimal Makefile                                    {#the-minimal-makefile}
+--------------------
+At minimum the Makefile of an application (see @ref getting-started) needs to
```
Comment (PeterKietzmann) [inline doc/doxygen/src/creating-an-application.md]: Needs to "**define**" the macros, right?

File: doc/doxygen/src/creating-an-application.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+`BOARD_INSUFFICIENT_MEMORY`. Any board name that is not included in
+`BOARD_WHITELIST` will issue a message that one has to expect errors if they
+build the application for the board referred by that name. The list can also be
+used by a CI system to not build the application for this board at all. A board
+that is included in `BOARD_BLACKLIST` will show the same behavior. The build
+system evaluates `BOARD_WHITELIST` first and then `BOARD_BLACKLIST`. The
+`BOARD_INSUFFICIENT_MEMORY` macro is similar to `BOARD_BLACKLIST`, but will
+build in any case. A CI system can use the information provided by the
+`BOARD_INSUFFICIENT_MEMORY` macro to skip the linking step in the build
+process, since some linkers will issue an error when the code won't fit the
```
Comment (PeterKietzmann) [inline doc/doxygen/src/creating-an-application.md]: Not sure but I think it should be "if" the code won't fit...

File: doc/doxygen/src/creating-an-application.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+RIOTBASE ?= $(CURDIR)/../../RIOT
+
+include $(RIOTBASE)/Makefile.include
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+How to handle unsupported boards?                  {#handle-unsupported-boards}
+---------------------------------
+Sometimes it is necessary to exclude boards because they don't provide a
+required functionality or don't have sufficient memory. RIOT's build system
+looks for the macros `BOARD_BLACKLIST`, `BOARD_WHITELIST`, and
```
Comment (PeterKietzmann) [inline doc/doxygen/src/creating-an-application.md]: Comma before "and" can be removed I guess

File: doc/doxygen/src/creating-an-application.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+`BOARD_INSUFFICIENT_MEMORY`. Any board name that is not included in
+`BOARD_WHITELIST` will issue a message that one has to expect errors if they
+build the application for the board referred by that name. The list can also be
+used by a CI system to not build the application for this board at all. A board
+that is included in `BOARD_BLACKLIST` will show the same behavior. The build
+system evaluates `BOARD_WHITELIST` first and then `BOARD_BLACKLIST`. The
+`BOARD_INSUFFICIENT_MEMORY` macro is similar to `BOARD_BLACKLIST`, but will
```
Comment (PeterKietzmann) [inline doc/doxygen/src/creating-an-application.md]: Comma before "but" can also be removed, doesn't it?

**Comment by PeterKietzmann** [comment]
Besides the minor comments: I like the your documentation enhancement

File: doc/doxygen/src/creating-an-application.md (hunk 15 above)
Comment (authmillenon) [inline doc/doxygen/src/creating-an-application.md]: Nope, I prefer the [serial comma](https://en.wikipedia.org/wiki/Serial_comma)

**Comment by authmillenon** [comment]
Addressed comments

**Comment by PeterKietzmann** [comment]
Well, ACK (when squashed) from my side. @kaspar030 might also want to check and push the merge button then

**Comment by authmillenon** [comment]
Squashed

**Comment by authmillenon** [comment]
Static tests pass, other stuff is not touched => go.

**Pull request merged**